packages/block-editor/src/hooks/auto-inserting-blocks.js

```diff
@@ -0,0 +1,214 @@
+
+/**
+ * Internal dependencies
+ */
+import { InspectorControls } from '../components';
+import { store as blockEditorStore } from '../store';
+
+function AutoInsertingBlocksControl( props ) {
+	const autoInsertedBlocksForCurrentBlock = useSelect(
+		( select ) => {
+			const { getBlockTypes } = select( blocksStore );
+			return getBlockTypes()?.filter(
+				( block ) =>
+					block.autoInsert &&
+					Object.keys( block.autoInsert ).includes( props.blockName )
+			);
+		},
+		[ props.blockName ]
+	);
+
+	const {
+		autoInsertedBlockClientIds,
+		blockIndex,
+		rootClientId,
+		innerBlocksLength,
+	} = useSelect(
+		( select ) => {
+			const { getBlock, getBlockIndex, getBlockRootClientId } =
+				select( blockEditorStore );
+			const _rootClientId = getBlockRootClientId( props.clientId );
+
+			const _autoInsertedBlockClientIds =
+				autoInsertedBlocksForCurrentBlock.reduce( ( acc, block ) => {
+					const relativePosition =
+						block?.autoInsert?.[ props.blockName ];
+					let autoInsertedBlock;
+					if ( [ 'before', 'after' ].includes( relativePosition ) ) {
+						const siblings = getBlock( _rootClientId )?.innerBlocks;
+
+						// Any of the current block's siblings (with the right block type) qualifies
+						// as an auto-inserted block (inserted `before` or `after` the current one),
+						// as the block might've been auto-inserted and then moved around a bit by the user.
+						autoInsertedBlock = siblings.find(
+							( { name } ) => name === block.name
+						);
+					} else if (
+						[ 'first_child', 'last_child' ].includes(
+							relativePosition
+						)
+					) {
+						const { innerBlocks } = getBlock( props.clientId );
+
+						// Any of the current block's child blocks (with the right block type) qualifies
+						// as an auto-inserted first or last child block, as the block might've been
+						// auto-inserted and then moved around a bit by the user.
+						autoInsertedBlock = innerBlocks.find(
+							( { name } ) => name === block.name
+						);
+					}
+
+					if ( autoInsertedBlock ) {
+						acc[ block.name ] = autoInsertedBlock.clientId;
+					}
+
+					// TOOD: If no auto-inserted block was found in any of its designated locations,
+					// we want to check if it's present elsewhere in the block tree.
+					// If it is, we'd consider it manually inserted and would want to remove the
+					// corresponding toggle from the block inspector panel.
+
+					return acc;
+				}, {} );
+
+			return {
+				blockIndex: getBlockIndex( props.clientId ),
+				innerBlocksLength: getBlock( props.clientId )?.innerBlocks
+					?.length,
+				rootClientId: _rootClientId,
+				autoInsertedBlockClientIds: _autoInsertedBlockClientIds,
+			};
+		},
+		[ autoInsertedBlocksForCurrentBlock, props.blockName, props.clientId ]
+	);
+
+	const { insertBlock, removeBlock } = useDispatch( blockEditorStore );
+
+	if ( ! autoInsertedBlocksForCurrentBlock.length ) {
+		return null;
+	}
```

Comment:
can this be moved above the `useSelect` and `useDispatch` calls or would that break the rule of hooks? if so, maybe it would skip a lot of store exchanges that we know aren't necessary

Reply:
> or would that break the rule of hooks?

That's what I thought (and why I didn't try moving it further up, which would of course be beneficial). I can give it a quick try to see if I'm mistaken.

Reply:
Yeah, linter says no 😬 

(BTW `autoInsertedBlocksForCurrentBlock` is computed inside the `useSelect` hook call anyway, so we wouldn't be able to cut out _that_ much by moving the early return conditional further up.)